Bottom: 23d1b8ac262ad8c25aa048ce092d0a3bf43a14cf
Top:    3b26b695382c0e9b77daf86b3c7c8f0e757f3e62
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2015-02-15 22:35:13 +0000

Name polygon edges separately (not contours), use these to name STEP faces

XXX: Incomplete.. needs us to actually set the name flags on the contour edges!


---

diff --git a/src/hid/common/object3d.c b/src/hid/common/object3d.c
index 90941d0..3ca032b 100644
--- a/src/hid/common/object3d.c
+++ b/src/hid/common/object3d.c
@@ -276,6 +276,22 @@ typedef struct
   face3d *bottom_face;
 } polygon_3d_link;
 
+
+/* NB: contour edge 'n' starts at contour coord 'n', and ends at coord 'n + 1' */
+static char *
+get_contour_edge_n_name (PLINE *contour, int n)
+{
+  VNODE *vertex = &contour->head;
+
+  while (n > 0) {
+    vertex = vertex->next; /* The VNODE structure is circularly linked, so wrapping is OK */
+    n--;
+  }
+
+  return vertex->edge_name; /* NB: Vertex 'n' is used to store the name for edge 'n' */
+}
+
+
 /* NOTE: This function sets the user_data pointer on POLYAREA it
  *       converts, to point at a polygon_3d_link structure
  *       referencing the generated object and upper/lower faces
@@ -399,10 +415,27 @@ object3d_from_contours (POLYAREA *contours,
           object3d_add_edge (object, edges[i]);
         }
 
+      ct = outer_contour;
+      offset_in_ct = 0;
+      ct_npoints = get_contour_npoints (ct);
+
       /* Define the faces */
-      for (i = 0; i < npoints; i++)
+      for (i = 0; i < npoints; i++, offset_in_ct++)
         {
-          faces[i] = make_face3d (NULL);
+          char *face_name;
+
+          /* Update which contour we're looking at */
+          if (offset_in_ct == ct_npoints)
+            {
+              offset_in_ct = 0;
+              ct = ct->next;
+              ct_npoints = get_contour_npoints (ct);
+            }
+
+          // face_name = (ct->name != NULL) ? ct->name : ""; // Naming all faces of a contour with the contour name
+          face_name = get_contour_edge_n_name (ct, adjusted_i); // Naming each face of a contour separately. (XXX: Should we prepend the contour name?)
+
+          faces[i] = make_face3d (face_name);
 
           object3d_add_face (object, faces[i]);
           /* Pick one of the upright edges which is within this face outer contour loop, and link it to the face */
diff --git a/src/polyarea.h b/src/polyarea.h
index acdbf1f..867e4c7 100644
--- a/src/polyarea.h
+++ b/src/polyarea.h
@@ -94,6 +94,7 @@ struct VNODE
     bool is_round;
     Coord cx, cy;
     Coord radius;
+    char *edge_name;
 };
 
 struct PLINE
diff --git a/src/polygon1.c b/src/polygon1.c
index 3ea0ebf..d7adeb7 100644
--- a/src/polygon1.c
+++ b/src/polygon1.c
@@ -3697,6 +3697,8 @@ poly_DelContour (PLINE ** c)
       else
         free (cur->cvc_prev);
 
+      free (cur->edge_name);
+
       g_slice_free (VNODE, cur);
     }
